# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(50)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   
30         1  335581.668515         62              среднее             1   
31         0   -1682.083438         47            начальное             3   
32         0   -4649.910832         34              СРЕДНЕЕ             1   
33         0   -1548.637544         48              среднее             1   
34         0   -4488.067031         35              среднее             1   
35         0  394021.072184         68              среднее             1   
36         0    -176.216688         33              среднее             1   
37         0   -6448.810860         43               ВЫСШЕЕ             0   
38         0    -597.881827         25               высшее             0   
39         0    -650.587796         31               высшее             0   
40         0   -1030.362606         30              среднее             1   
41         0            NaN         50              среднее             1   
42         0   -1257.496190         20  неоконченное высшее             2   
43         0   -4375.681384         43  неоконченное высшее             2   
44         1   -1362.041728         26              СРЕДНЕЕ             1   
45         1   -1039.451741         49              среднее             1   
46         2   -2262.712304         37              среднее             1   
47         1   -2689.137274         33               высшее             0   
48         0   -3341.067886         45              среднее             1   
49         0   -1181.443228         54              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0  

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

In [9]:
len(data[data['children']==-1])


47

In [10]:
len(data[data['children']==20])

76

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [11]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [12]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [13]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [14]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [15]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [16]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [17]:
data.duplicated().sum()

71

In [18]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [19]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [34]:
def categorize_income_2(income):
    try:
        if 0 <= income <= 50000:
            return 'E'
        elif 50001 <= income <= 100000:
            return 'D'
        elif 100001 <= income <= 150000:
            return 'C'
        elif 150001 <= income <= 300000:
            return 'B'
        elif income >= 300001:
            return 'A'
    except:
        pass

In [20]:
data['total_income_category'] = data['total_income'].apply(categorize_income)


In [35]:
data['total_income_category_2'] = data['total_income'].apply(categorize_income_2)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [ ]:
data['purpose'].unique()

In [21]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [22]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [23]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [38]:
data['debt'].mean()

0.08119638085415593

In [24]:
aggfunc_=['sum','count','mean']#список который применится к сводной таблице.
data.pivot_table(index='children', values='debt', aggfunc=aggfunc_)


#data['children'].corr(data['debt'])
# Ваш код будет здесь. Вы можете создавать новые ячейки.

sum  count      mean
          debt   debt      debt
children                       
0         1063  14091  0.075438
1          444   4808  0.092346
2          194   2052  0.094542
3           27    330  0.081818
4            4     41  0.097561
5            0      9  0.000000

**Вывод:** 



Доля должников по группа отличается на 2.2% от минимального в группе у кого детей нет, до максимального у кого четыре ребенка(*не учитывается у кого пять детей из за малого количества в выборке*)
Зависимость между кол-во детей, для тех клиентов у которых уже есть дети очень слабая.Для оценки доли должников можно разделить по группам: "нет детей"(7,54%),"есть дети(9,24%)", так как между ними существует разница в 1,7 %.
p.s В индивидуальном порядке можно рассмотреть 9 клиентов у которых 5 детей и нет задолжености, которые для статистики находятся в области погрешности, но могут быть интересны для качественого иследования или маркетинга.
p.s.s Можно еще изучить снижение доли должников, для тех у кого трое детей, относительно одного-двух детей.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
aggfunc_=['sum','count','mean']#список который применится к сводной таблице.
data.pivot_table(index='family_status', values='debt', aggfunc=aggfunc_)
#data['family_status_id'].corr(data['debt'])
# Ваш код будет здесь. Вы можете создавать новые ячейки.

sum  count      mean
                      debt   debt      debt
family_status                              
Не женат / не замужем  273   2796  0.097639
в разводе               84   1189  0.070648
вдовец / вдова          63    951  0.066246
гражданский брак       385   4134  0.093130
женат / замужем        927  12261  0.075606

**Вывод:** 

Доля должников по группа отличается на 2.7% от максимальной для тех кто *Не женат/не замужем* и минимальной для группы *вдовец / вдова*.
Близкие значения показывают группы:  
1)разводе,вдовец/вдова и женат/замужем
2)Не женат/не замужем,гражданский брак


In [41]:
aggfunc_=['sum','count','mean']#список который применится к сводной таблице.
data.pivot_table(index='gender', values='debt', aggfunc=aggfunc_)#я решил еще посчитать для пола для лучшего вывода

sum  count      mean
       debt   debt      debt
gender                      
F       989  14092  0.070182
M       743   7238  0.102653
XNA       0      1  0.000000

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
aggfunc_=['sum','count','mean']#список который применится к сводной таблице.
data.pivot_table(index='total_income_category', values='debt', aggfunc=aggfunc_)# Ваш код будет здесь. Вы можете создавать новые ячейки.


sum  count      mean
                       debt   debt      debt
total_income_category                       
A                         2     25  0.080000
B                       354   5014  0.070602
C                      1353  15921  0.084982
D                        21    349  0.060172
E                         2     22  0.090909

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [36]:
aggfunc_=['sum','count','mean']#список который применится к сводной таблице.
data.pivot_table(index='total_income_category_2', values='debt', aggfunc=aggfunc_)# Ваш код будет здесь. Вы можете создавать новые ячейки.


sum count      mean
                        debt  debt      debt
total_income_category_2                     
A                        106  1474  0.071913
B                        653  8303  0.078646
C                        619  7110  0.087060
D                        331  4073  0.081267
E                         23   371  0.061995

- 0–50000 — `'E'`;
- 50001–100000 — `'D'`;
- 100001–150001 — `'C'`;
- 150001–300000 — `'B'`;
- 300001 и выше — `'A'`.



**Вывод:** 

Доля должников по группам отличается на 3,1% от максимальной для тех кто попадает в категорию Е (0–30000) и минимальной для категории D (30001–50000)
Возможно,лучше измениеть параметры категорий, так как сейчас распределение по ним не равномерно. 
upd Если разбить на другие группы (total_income_category_2), то можно заметить, что наибольшая доля должников в группе С, со среднем уровнем дохода.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [40]:
aggfunc_=['sum','count','mean']#список который применится к сводной таблице.
data.pivot_table(index='purpose_category', values='debt', aggfunc=aggfunc_)# Ваш код будет здесь. Вы можете создавать новые ячейки.
# Ваш код будет здесь. Вы можете создавать новые ячейки.

sum  count      mean
                         debt   debt      debt
purpose_category                              
операции с автомобилем    400   4279  0.093480
операции с недвижимостью  780  10751  0.072551
получение образования     369   3988  0.092528
проведение свадьбы        183   2313  0.079118

**Вывод:** 

Доля должников по группам отличается на 2,1% от максимальной для тех кто берет кредит на операции с автомобилем и минимальной для тех кто берет кредит на операции с недвижимостью
. 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* пропуски были в двух столбцах в одинаковом количестве  общий трудовой стаж и ежемесячный доход для одних и тех же клиентах, скорее всего эти данные были не объязательные и их не получили.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* В данном случае мы заполняли медианным значением ежемесячный доход, среднее значение могло быть сильно смещано из небольшого процента клиентов у которых самый большой доход, поэтому лучше было воспользоваться наиболее часто встречаемым значением.

### Шаг 4: общий вывод.

В ходе работы над проектом, мы ознакомились с данными  о 21525 клиентах банка , в которых были пропуски, скорее всего из-за того, что эти данные были необъязательны для ввода и клиненты их не указали. Трудовой стаж и ежемесечный доход мы заполнили медиаными значениями исходя из типа занятости. Так же в трудовом стаже, скорее всего, из-за технической ошибки были отрицательные значения, которую мы исправили.

Так же мы зафиксировали,что у безработных и пенсионерах аномально большие значения трудового стажа примерно 1000 лет, что аномально для людей, причина не понятна, исправить эту аномалию мы не стали,так как данные нам не были нужны в иследование, но стоит помнить о данной ошибке. В столбце с количеством детей были значения -1 и 20, вероятно, это опечатка,так как это 0,6% от количества значений и ,вероятно, можно было заменить их на 1 и 2 соответсвенно, но мы решили их удалить, что бы они не исказили данные иследования.

Мы создали функцию для категоризации по ключивым словам для разных целей кредита. 
Затем мы приступили к определения доли должников в зависимости от семейного положению, дохода. кол-во детей и цели кредита.
Средняя доля клиентов которые имеют  задолженность по возврату кредитов 8,1%
Клиенты у кого есть дети (9,2%)(7240 чел.) на 1,7% чаше имеют задолженость по кредиту, чем те у кого нет детей(7,5%)(14091 чел), разница среди тех у кого один или двое детей практически отсутсвует, родители троих детей на 1,2 % реже уходят в просрочку, чем родители двух детей, в выборке слишком мало клиентов у которых четверо и пятеро детей из за чего может возникать аномальные значения. 

Большая часть клиентов женаты/замужем(12261 чел.). По резултатам иследования в группу со схожей долей должников можно отнести:
1)Не женат/не замужем,гражданский брак (6930 чел.)
2)Женат/замужем, в разводе,вдовец/вдова  (25001 чел.)
Которые в среднем имеют долю должников в 9,5% и 7,3% соответсвенно.
В зависимости от уровня доходов, сначала мы поделили клиентов на 5 категорий:
- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

В категорию С попала около 74%, а в крайние групы 0,12 % - недостаточный размер для построения по ним надежых выводов.
Изменив классификацию, на:
- 0–50000 — `'E'`;
- 50001–100000 — `'D'`;
- 100001–150001 — `'C'`;
- 150001–300000 — `'B'`;
- 300001 и выше — `'A'`.

Удалось добится более равномерного распределения клиентов по группа и сделать вывод, что наибольшая доля должников (8,7%) в группе С со средним зароботоком в 100001–150001, доля должников уменьшается ,как в сторону группы с минимальным знаечением,группа Е (6,2 %) с доходом менее 50000, что вызывает вопросы, так и в сторону группы А (7,2%) с доходом больше 300001 

Разделение по цели кредита показала, что практически половина от всех кредитов выдано на опереации с недвижемостью, они же вместе с кредитами на свадьбу являются самыми безопасными для банка по доли просроченой задолженость 7,3 и 7,9% соответственно,противоположность им группы кредитов выданые для получения образования и операции с автомобилем  с долей должников 9,3% 

Обобщая можно сказать, что данные подходят для иследования с некоторой осторожностью,особенно по столбцу с трудовым стажем,а статистически значемы ли эти различия мы узнаем в следующих сериях нашего обучения.